docs/specification/draft/basic/authorization.mdx

```diff
@@ -263,35 +253,50 @@ Servers **MUST** return appropriate HTTP status codes for authorization errors:
 | 400         | Bad Request  | Malformed authorization request            |
 
-### 2.10 Implementation Requirements
+## 3. Security Considerations
+
+Implementations **MUST** follow OAuth 2.1 security best practices. Refer to
+[RFC9700](https://datatracker.ietf.org/doc/html/rfc9700) for details.
+
+### 3.1 Token Theft
+Attackers who obtain tokens stored by the client, by accessing tokens cached or logged on the server can access protected resources with
+requests that appear legitimate to resource servers.
+
+Clients **MUST** implement secure token storage and follow OAuth 2.0 best practices,
+as outlined in [RFC 9700](https://datatracker.ietf.org/doc/html/rfc9700).
+
+MCP authorization servers SHOULD enforce token expiration and rotation to limit the window of exploitation.
+
+### 3.2 Token Interception
+An attacker positioned between MCP clients and MCP servers can intercept tokens via [Man-in-the-Middle (MITM)](https://en.wikipedia.org/wiki/Man-in-the-middle_attack) attacks.
+
+To mitigate the risk of this threat:
+
+1. All authorization endpoints **MUST** be served over HTTPS.
+1. MCP clients **MUST** implement PKCE according to [OAuth 2.1 section 7.5.2](https://www.ietf.org/archive/id/draft-ietf-oauth-v2-1-12.html#name-countermeasures). PKCE helps  prevent authorization code interception attacks by requiring clients to create a secret verifier-challenge pair, ensuring that only the original requestor can exchange an authorization code for tokens.
+1. All redirect URIs **MUST** be either `localhost` or use HTTPS to prevent token and code interception.
+
+### 3.3 Open Redirection
+An attacker may craft malicious redirect URIs to direct users to phishing sites 
+and intercept credentials during the authorization flow.
 
-1. Implementations **MUST** follow OAuth 2.1 security best practices
-1. PKCE is **REQUIRED** for all MCP clients and authorization servers
-1. MCP servers that also act as an AS:
-    1. **SHOULD** implement token rotation for enhanced security
-    1. **SHOULD** restrict token lifetimes based on security requirements
+MCP clients **MUST** have redirect URIs registered with the authorization server.
 
-## 3. Best Practices
+Authorization servers **MUST** validate exact redirect URIs against pre-registered values to prevent redirection attacks.
 
-#### 3.1 Local clients as Public OAuth 2.1 Clients
+MCP clients **SHOULD** use and verify state parameters in the authorization code flow
+and discard any results that do not include or have a mis-match with the original state.
 
-We strongly recommend that local clients implement OAuth 2.1 as a public client:
 
-1. Utilizing code challenges (PKCE) for authorization requests to prevent interception
-   attacks
-2. Implementing secure token storage appropriate for the local system
-3. Following token refresh best practices to maintain sessions
-4. Properly handling token expiration and renewal
+### 3.4 Confused Deputy Problem
```

Comment:
This section needs some cleanup on the terminology used. There's a few different terms being used to refer to the same things or aren't previously defined:

* OAuth proxy
* "...fronts an authorization server"
* "backing authorization server"
* "upstream API"

We need to be much clearer about which roles we're talking about here. I would recommend starting with an abbreviated sequence diagram to show the relationships between the different parties, as well as a definition of each term, prior to discussing the attack.

Reply:
👍 I'm going to split this out into it's own PR since I think there will be some back and forth to get it right.  This PR moves some other sections around so want to get it in to avoid merge conflicts.

https://github.com/modelcontextprotocol/modelcontextprotocol/pull/407